File: frameworks/test.persistence/src/main/java/com/everest/frameworks/test/persistence/Instances.java
package com.everest.frameworks.test.persistence;

import com.everest.frameworks.domain.core.Exceptions;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Field;
import java.util.Arrays;
import java.util.List;

public class Instances {

    private final List<Object> instances;

    public Instances(List<Object> instances) {
        this.instances = instances;
    }

    public void autoAssignPersistence() {
        this.instances.forEach(this::autoAssignPersistence);
    }

    private void autoAssignPersistence(Object instance) {
        Class<?> instanceClass = instance.getClass();
        Field[] fields = instanceClass.getDeclaredFields();
        Arrays.stream(fields).forEach(field -> {
            if (!field.isAnnotationPresent(Persistence.class)) return;
            field.setAccessible(true);
            Exceptions.execute(() -> field.set(instance, Persistences.fetchByType(type(field))));
        });
    }

    @NotNull
    protected Class<?> type(Field field) {
        Class<?> type = field.getType();
        if (!Persistences.exist(type)) throw new PersistenceException("Error! The type is not register in Persistence:" + type.getName()
                + ". You need to check if the type is valid. If it is a new persistence template, you can add it to Persistence in " +
                "frameworks.test.persistence.");
        return type;
    }

}

File: frameworks/test.persistence/src/main/kotlin/com/qomolangma/frameworks/test/persistence/Instances.kt
package com.qomolangma.frameworks.test.persistence

import com.qomolangma.frameworks.domain.core.Exceptions.execute
import java.util.*

class Instances(private val instances: List<Any>) {
    fun autoAssignPersistence() = instances.forEach { autoAssignPersistence(it) }

    private fun autoAssignPersistence(instance: Any) {
        val fields = instance.javaClass.declaredFields
        Arrays.stream(fields).forEach {
            if (!Persistences.exist(it.type)) return@forEach
            it.isAccessible = true
            execute { it[instance] = Persistences.fetchByType(it.type) }
        }
    }
}
